feat: transform plugins

zsilbi opened the pull request:
Resolves #6
Resolves #33 
Resolves #35 

> Sure ideally we can start with an mkdist-compatible context API for custom extensions in transpiler. 

 _Originally posted by @pi0 in [#6](https://github.com/unjs/obuild/issues/6#issuecomment-2913429285)_
 
I made this as a starting point for further discussions about #6, it's mainly based on the `mkdist` loaders.

### New features in this PR:
#### Plugin support for transform entries. 
Plugin monorepo will be available here soon: [https://github.com/unjs/obuild-plugins](https://github.com/unjs/obuild-plugins)
Until the packages are not published, they are packed and resolved from the `/plugins` dir in this PR.

Possible setup:
```ts
 {
    type: "transform",
    input: "src/",
    outDir: "dist/",
    plugins: [
      oxcDts({ force: true }), // Force generate, skip emitting a file if isolatedDeclarations fails
      tsgoDts(), // Reuse .d.(m)ts from `oxc-dts` and generate for TS files that `oxc-dts` could not work on
      vueTscDts(), // Reuse .d.(m)ts from `oxc-dts` and `tsgo-dts` and generate only for Vue files
      vueSFCTransformer(), // Transform Vue SFCs
      oxcTransform(), // Transform others
      sass(),
      postcss(),
    ],
  }
 ```

Plugins have a `transform()` method, we call these recursively in the order that is provided in the configuration.
- if it returns  with an array of files, we go through each file and try to call the next plugins transform method with it
- if it returns undefined we go to the next plugin

Plugins can also have temp. storage assigned to them for each entry context. This is useful when you for example define the plugins  once and use them for multiple entries with the same configuration.

#### TS Config resolution
On build start we find and resolve all paths in the TS Config to absolute paths based on where the actual configuration coming from. (TS config entries can also be added into the transform entries).
Later the transform plugins can use these to make relative paths based on their requirements.

#### Source map serialization/path rewrite
The plugins don't have enough context to know where each file will be written to, this is done here.
Added `mapDir` config option to write these separately. (This directory is not cleaned on each build automatically)

#### `declarationDir` support
When provided in TS config, the declarations are written in the provided directory (This directory is not cleaned on each build automatically)

zsilbi committed:
feat(transform): add resolve options for `TransformEntry`

zsilbi committed:
feat(transform): use default resolve options

zsilbi committed:
feat: allow passing complete exsolve options

zsilbi committed:
update resolve options

zsilbi committed:
feat(tranform): extendable transformers

zsilbi committed:
feat(transformer): jsx/tsx support

zsilbi committed:
docs: update readme

zsilbi committed:
fix: types

zsilbi committed:
fix: oxc transform config

zsilbi committed:
provided transformers have higher priority

zsilbi committed:
chore: seperate tsconfig for test (w/o isolatedDeclarations)

zsilbi committed:
test: reogranize fixtures

zsilbi committed:
chore: update tsconfig for tests

zsilbi committed:
chore: install dependencies for vue transformer

zsilbi committed:
fix: oxc non-transform output

zsilbi committed:
feat: use vue-sfc-transformer/mkdist as vue transformer

zsilbi committed:
refactor: improve readability

autofix-ci[bot] committed:
chore: apply automated updates

zsilbi committed:
fix: js extensions

zsilbi committed:
chore: clarify types

pi0 commented:
Cool! We probably have to also support alternative `dts` transformer based on this later for (non infered) type gen.

zsilbi commented:
> Cool! We probably have to also support alternative `dts` transformer based on this later for (non infered) type gen.

I'll leavse some notes here about current process if we want to do this later.

With `mkdist` the Vue declarations are generated in the following way:

1. `vue-sfc-transformer/mkdist` transforms them as `.js` files
https://github.com/nuxt-contrib/vue-sfc-transformer/blob/483fbf466332ef97d9ec3dbb58c9f2c6d4fa03d7/src/utils/mkdist.ts#L87-L92

2. `mkdist` adds declaration outputs for every `.js` files (why not only `.ts`?) in the `js` loader
https://github.com/unjs/mkdist/blob/a7f029d9dc4ee083f7426d383f1622df4927eb09/src/loaders/js.ts#L22-L33

3. `vue-sfc-transformer/mkdist` filters these added delarations outputs and adds them to it's own outputs
https://github.com/nuxt-contrib/vue-sfc-transformer/blob/483fbf466332ef97d9ec3dbb58c9f2c6d4fa03d7/src/utils/mkdist.ts#L93

4. `mkdist` later processes all declarations files using two utility loaders
https://github.com/unjs/mkdist/blob/a7f029d9dc4ee083f7426d383f1622df4927eb09/src/make.ts#L111-L126

pi0 commented:
Thanks for the notes.

/cc @danielroe @Teages re vue-sfc types part.

zsilbi committed:
refactor: move transformers dir, update types and context

zsilbi committed:
chore: add /transformers subpath export

zsilbi committed:
refactor: remove generic options

zsilbi committed:
chore: enable "verbatimModuleSyntax" in tsconfig

zsilbi committed:
fix: type-only imports

zsilbi committed:
docs: update readme with transformers

zsilbi committed:
docs: correct examples

Teages commented:
i think we can have the fallback transformer as mkdist did, since vue-tsc is not support allowArbitraryExtensions yet.

zsilbi commented:
> i think we can have the fallback transformer as mkdist did, since vue-tsc is not support allowArbitraryExtensions yet.

We should address the vue type emit seperatetely and add warning about the missing types in this PR.

zsilbi committed:
fix: remove unused config type

zsilbi committed:
refactor: improved mkdist compatibilty

zsilbi committed:
chore: rename variable

zsilbi committed:
chore: update error message

zsilbi committed:
chore: fix typo in comment

zsilbi committed:
fix: pass mkdist options to loader context

zsilbi committed:
chore: update warning message

zsilbi committed:
chore: update warning message

zsilbi committed:
fix: remove unnecessary async

zsilbi committed:
fix: remove unecessary context merge

zsilbi committed:
fix: pass genereate boolean more accurately

zsilbi committed:
docs: update readme

zsilbi committed:
docs: update readme

pi0 committed:
minor refactors

pi0 committed:
simplify tests

zsilbi committed:
fix: type import

zsilbi committed:
fix: mkdist conversion

zsilbi committed:
fix: transformer type exports

zsilbi committed:
refactor: extract minifer from oxc transformer

zsilbi committed:
fix: remove redundant extension property

zsilbi committed:
fix: unset skip property in minified output

zsilbi committed:
chore: disable verbatimModuleSyntax for test files

zsilbi committed:
test: add test files for source maps

zsilbi committed:
refactor: oxc transformer

zsilbi committed:
refactor: oxc transformer

autofix-ci[bot] committed:
chore: apply automated updates

zsilbi committed:
fix: correct type definitions

zsilbi committed:
test: monitor dist output snapshot

zsilbi committed:
fix: source map paths

zsilbi committed:
update mkdist compat

zsilbi committed:
feat: declarations generation

zsilbi committed:
test: update snapshots

zsilbi committed:
docs: remove notice

zsilbi committed:
test: update snapshots

zsilbi committed:
chore: update deps

zsilbi committed:
chore: remove external from config

zsilbi committed:
fix: oxc transformer skip declaration files

zsilbi committed:
fix: oxc use correct target extensions, support .mts

zsilbi committed:
test: update output tests

zsilbi committed:
test: update snapshot

zsilbi committed:
fix: declaration check regex

autofix-ci[bot] committed:
chore: apply automated updates

zsilbi committed:
fix: enable jsx type declarations

zsilbi committed:
fix: don't merge default transformers with config

zsilbi committed:
docs: add todo

zsilbi committed:
fix: dont rename js

zsilbi committed:
test: vue transformer

zsilbi committed:
fix: vue transformer caching

zsilbi committed:
refactor: simplify file types

zsilbi committed:
Merge branch 'main'

zsilbi committed:
test: increase timeout

zsilbi committed:
test: update timeout option

zsilbi commented:
I'll add more tests to increase coverage.
Support for older `vue-tsc` versions is also included, do we have want to keep that?

zsilbi committed:
feat: drop legacy vue-tsc support

zsilbi committed:
feat: declaration options

zsilbi committed:
fix: force isolatedDeclarations for vue-tsc

zsilbi committed:
feat: postcss transformer

zsilbi committed:
feat: sass transformer

zsilbi committed:
test: vue scss

zsilbi committed:
docs: add description for sass transformer

zsilbi committed:
fix: update `TransformerName` with the new transformers

zsilbi committed:
fix: simplify extensions

zsilbi committed:
fix: add decl. type to declaration files

zsilbi committed:
test: dts only custom transformer

zsilbi committed:
test: update snapshots

zsilbi committed:
test: update timeout

zsilbi committed:
fix: dts output for .d.mts extensions

zsilbi committed:
test: dts dir import

zsilbi committed:
test: dts dynamic imports

zsilbi committed:
docs: update readme

autofix-ci[bot] committed:
chore: apply automated updates

zsilbi committed:
docs: update readme

zsilbi committed:
feat: restore warning for missing `vue-tsc`  dep.

zsilbi committed:
refactor: shebang check

zsilbi committed:
fix: set unchanged files raw

zsilbi committed:
feat: stream copy raw files with shebang check

zsilbi committed:
refactor: remove unnecessary array check and prop

zsilbi committed:
chore: replace node:path with pathe

zsilbi committed:
refactor: file copy with shebang check

zsilbi committed:
fix: improve raw files handling

zsilbi committed:
refactor: simplify file types

zsilbi committed:
fix: file types

zsilbi committed:
chore: revert tsconfig split

zsilbi committed:
chore: turn verbatimModuleSyntax back on

zsilbi committed:
fix: remove duplicate console.log for dts error check

zsilbi committed:
feat: combined (minifed -> ts) source maps

zsilbi committed:
refactor: split oxc transformer

zsilbi committed:
refactor: rename types

zsilbi committed:
refactor: source map merge

zsilbi committed:
refactor: minor updates

zsilbi committed:
refactor: improve type safety

zsilbi committed:
perf: swap conditions

zsilbi committed:
refactor: rename type

zsilbi committed:
fix. source map merge input types

zsilbi committed:
chore: rearrange imports

zsilbi committed:
refactor: oxc transform tuples

zsilbi committed:
fix: remove uneeded spread

zsilbi committed:
fix: minify options

zsilbi committed:
fix: transform options

zsilbi committed:
refactor: simplify option types

zsilbi committed:
docs: add references to mkdist sources

zsilbi committed:
refactor: simplify options

zsilbi committed:
perf: swap conditions

zsilbi committed:
refactor: source config

zsilbi committed:
refactor: split transformDir

zsilbi committed:
docs: add tsdoc to methods

zsilbi committed:
feat: `mapDir` option for transform entries

zsilbi committed:
refactor: rename utils to declarations

zsilbi committed:
test: update snapshot with correct paths

zsilbi committed:
feat: non-isolated declarations emit from oxc transformer

zsilbi committed:
refactor: normalize mapDir

zsilbi committed:
docs: update tsdoc

zsilbi committed:
refactor: createTransformer

zsilbi committed:
fix: use single new lines in console

zsilbi committed:
feat: read nested tsconfig

zsilbi committed:
docs: clarfiy oxc dts fallback

zsilbi committed:
fix: remove unused option

zsilbi committed:
fix: export  compat loader options type

zsilbi committed:
docs: remove note

zsilbi committed:
refactor: move dts disable in transform

zsilbi committed:
feat: tsgo (WIP)

zsilbi committed:
test: refactor test suite (WIP)

zsilbi committed:
test: refactor test suite

zsilbi committed:
feat: tsgo

zsilbi committed:
chore: update .gitignore to ignore all tsbuildinfo files

zsilbi committed:
chore: split tsconfig file to disable isolated decl. in tests

zsilbi committed:
build: fix type check

zsilbi committed:
test: refactor test suite

zsilbi committed:
test: relative extensions

zsilbi committed:
fix: relative extensions

zsilbi committed:
chore: revert type testing to src

zsilbi committed:
docs: add func. tsdoc

zsilbi committed:
test: increase timeout

zsilbi committed:
feat: rolldown-dts tsgo support

zsilbi committed:
refactor: allow setting tsgo option per entry

zsilbi committed:
fix: remove unused option

zsilbi committed:
docs: add `tsgo` option to readme

Teages commented:
> If that really solves it, why isn't that ported to `vue-sfc-transformer`? Thanks!

see https://github.com/nuxt/module-builder/issues/597 https://github.com/vuejs/core/issues/13295

In previous versions, Vue only load types from '.vue' file itself instead of its declarations (`.d.vue.ts` or `.vue.d.ts`), so when the Vue SFC export some type, it may cause error.

-----------

For example:

in `a.vue`
```vue
<script lang="ts">
export interface AProps { message: string }
</script>
```

When I want to extend props of `a.vue` in `b.vue`
```vue
<script lang="ts">
export interface BProps extends AProps { another: number }
//                              ^ the type is in `a.vue.d.ts` or `a.d.vue.ts`
//                                but vue can't find it because it only load the `a.vue` file
</script>

<script setup lang="ts">
const props = defineProps<BProps>()
</script>
```
-----------

Of course, this is all optional now. [This PR](https://github.com/vuejs/core/pull/13301) fixed it and was recently released in Vue 3.5.15

zsilbi commented:
Thank you @Teages!

We can emit both versions of the declarations if that's enough for now. 

We will probably need `hooks` support for the `TransformerContext`.
My idea would be to allow any `Transformer` use hooks to alter the finalized output of the process. 

In this example the `vue` transformer would subscribe to the `declarationsGenerated()` hook and duplicate  the `.vue.d.ts` declaration output files with the new extension.

### tsgo

Building `nuxt/packages/nuxt` seems to work now with `tsgo`:
- unbuild: 9934 ms
- obuild + tsc: 9126 ms
- obuild + tsgo: 5140 ms

There are only 3 minor TS warnings about type portability in the transform entries.
Enabling `tsgo` for the bundled entry causes lots of TS errors so I could not test it with `rolldown-plugin-dts`.

The bundle size probably needs some investigation as it's a bit smaller with `tsgo`:
- unbuild: Σ Total dist size (byte size): 816 kB
- tsc: Σ Total dist byte size: 802 kB (275 files)
- tsgo: Σ Total dist byte size: 769 kB (275 files)

zsilbi committed:
fix: remove unnecessary check

zsilbi committed:
chore(deps): update dependencies

zsilbi commented:
> There are only 3 minor TS warnings about type portability in the transform entries. Enabling `tsgo` for the bundled entry causes lots of TS errors so I could not test it with `rolldown-plugin-dts`.
> 
> The bundle size probably needs some investigation as it's a bit smaller with `tsgo

I had the `noCheck` flag enabled for `tsgo`. Removing that reveals a lot more errors in the transform entries.
This can also explain bundle size decrease.

Should we also fail here when `tsgo` returns non `0` exit code?

zsilbi committed:
fix: remove `tsgo` noCheck

zsilbi committed:
fix: tsgo error handling

pi0 commented:
> Should we also fail here when tsgo returns non 0 exit code?

Perhaps we can show a warning in the end of build about possible issues? (or ⚠️ along those entries)

zsilbi commented:
> > Should we also fail here when tsgo returns non 0 exit code?
> 
> Perhaps we can show a warning in the end of build about possible issues? (or ⚠️ along those entries)

Currently I pipe the output from `tsgo`, that will display the formatted errors, then:

```bash
... 
src/validate.mts:4:53 - error TS7006: Parameter 'from' implicitly has an 'any' type.

export default defineNuxtRouteMiddleware(async (to, from) => {
                                                    ~~~~

Found 48 errors in 8 files.

Errors  Files
     2  src/composables.mts:6
     8  src/page.mts:9
     5  src/plugins/check-if-page-unused.mts:2
     6  src/plugins/prefetch.client.mts:3
     5  src/plugins/prerender.server.mts:6
    10  src/plugins/router.mts:7
     8  src/router.options.mts:3
     4  src/validate.mts:1

ℹ tsconfig.json:                                                                                                                                                                                                        
{
  compilerOptions: {
    allowJs: true,
    ...
    },
    ...
}

[WARN]  Error while generating declarations with tsgo: Process exited with code 2 
 ```
 
The source of these error is that the custom paths used in `tsconfig.json` are incorrectly resolved at the moment. I am working on a fix.

zsilbi committed:
fix: improve tsgo error debug

zsilbi committed:
refactor: dts types

zsilbi commented:
I think using `--noCheck` will be a required as these errors seem to have no effect on the output compared to `unbuild`.
In order to keep the `diff` minimal I had to turn off converting the files to `.mjs` files and used `.js` instead.

I created a version that uses `tsc` instead of `tsgo` but in the same exact way and compared the `*.d.ts` files generated:
[unbuild VFS  vs. obuild CLI tsc](https://gist.github.com/zsilbi/80544c4115328ae0c64ea23daa43abbe)

Then compared that with the `tsgo` version:
[obuild CLI tsc vs. obuild CLI tsgo](https://gist.github.com/zsilbi/522422d79fce8c342d42f1f49caf87be)

The diff is larger, but this `tsgo` run notified me about these issues (even with the added `--noCheck` added flag):
```bash
 WARN  The tsgo option is experimental and may change in the future.                                                                                                                                                                                                          3:02:14 PM

src/components.ts:128:14 - error TS2742: The inferred type of 'NoScript' cannot be named without a reference to '../../../home/zsilbi/projects/nuxt/node_modules/.pnpm/unhead@2.0.10/node_modules/unhead/dist/types.js'. This is likely not portable. A type annotation is necessary.

128 export const NoScript = defineComponent({
                 ~~~~~~~~
src/components.ts:157:14 - error TS2742: The inferred type of 'Link' cannot be named without a reference to '../../../home/zsilbi/projects/nuxt/node_modules/.pnpm/unhead@2.0.10/node_modules/unhead/dist/types.js'. This is likely not portable. A type annotation is necessary.

157 export const Link = defineComponent({
                 ~~~~
src/components.ts:268:14 - error TS2742: The inferred type of 'Style' cannot be named without a reference to '../../../home/zsilbi/projects/nuxt/node_modules/.pnpm/unhead@2.0.10/node_modules/unhead/dist/types.js'. This is likely not portable. A type annotation is necessary.

268 export const Style = defineComponent({
                 ~~~~~

Found 3 errors in the same file, starting at: src/components.ts:128
```

> The source of these error is that the custom paths used in `tsconfig.json` are incorrectly resolved at the moment. I am working on a fix.

Resolving these paths correctly had no positive effect, the errors were still reported and both `tsc` and `tsgo` decided to emit some of the declaration files n the original `src` dir.

zsilbi committed:
fix: revert tsconfig path resolve

zsilbi committed:
feat: add tsc to tsgo for debugging

zsilbi committed:
test: update snapshots with declaration comments

zsilbi committed:
fix: typo

zsilbi committed:
fix: transform result restruct

zsilbi commented:
> current PR seems to be little off track in latest state and probably we should down the scope and modulize development.
> 
> From obuild core i expect to have:
> 
> * pluggable interface for external transformers
> * oxc transform like main (.ts to .mjs and .d.mts)
> 
> Truesting the pluggable interface is tested now, we can init new monorepo you can manage for transformers (tsx, tsc, tsgo, vue and mkdist, …)
> 
> Sounds good?

Thanks @pi0, I agree! Let's do this!  🚀🚀

### tsgo and tsc (CLI)

> I created a version that uses `tsc` instead of `tsgo` but in the same exact way and compared the `*.d.ts` files generated: [unbuild VFS vs. obuild CLI tsc](https://gist.github.com/zsilbi/80544c4115328ae0c64ea23daa43abbe)

I updated the gists and removed:
 - `*.d.ts` files emitted by `rolldown-plugin-dts` as that's not relevant here
 - every entry where there the type looks identica
 
The main focus is on `unbuild vs obuild tsc CLI` (`tsc` and `tsgo` are supposed to be identical)

4 files left in the diff, all 4 of them are located in `/src/core/runtime/nitro/` 
One file is missing from the new build which is a `.js` file that has `// @ts-check` enabled. 

**Now I understand why `mkdist` adds declaration files for every `.js` file as well.**

### tsc (VFS)

While I was there I ran the `unbuild vs obuild tsc VFS` comparison, and there are [no missing types there](https://gist.github.com/zsilbi/df85795e7729fbd5c06330741f86434b), in fact there is improvement 👍🏻 

Tthis is the version that we should use as stable.

zsilbi commented:
> 4 files left in the diff, all 4 of them are located in `/src/core/runtime/nitro/` One declaration file is missing from the new build which is a `.js` file that has `// @ts-check` enabled.
> 
> **Now I understand why `mkdist` adds declaration files for every `.js` file as well.**

In order to support generating declaration files for `.js` files we have to process every file through `oxc-transform` to get declarations instantly if `isolatedDeclarations` is enabled. Rolldown does this by default also as I tested.

zsilbi committed:
fix: js files dts emit

zsilbi committed:
test: revert snapshot update

zsilbi committed:
fix: dts gen. condition

zsilbi committed:
test: update correct snapshot

zsilbi committed:
refactor: tsgo -> tsc-cli declarations

zsilbi committed:
docs: remove todo

zsilbi committed:
chore: update tsgo

zsilbi committed:
refactor: transform with dts tsc/tsgo cli

zsilbi committed:
chore: ignore .obuild folders

zsilbi committed:
test: remove providing tsconfig manually

zsilbi committed:
chore: reorganize imports

zsilbi committed:
refactor: tsc-cli

zsilbi committed:
feat: vue-tsc cli

zsilbi committed:
fix: revert default compiler to `tsgo`

zsilbi committed:
fix: use cross-platform symlinks

zsilbi committed:
refactor: move plugins to monorepo

zsilbi committed:
fix: ci errors

zsilbi committed:
chore: update plugins

pi0 commented on package.json:
```diff
@@ -27,18 +27,45 @@
+    "@obuild/plugin": "file:plugins/obuild-plugin-0.0.1.tgz",
+    "@obuild/plugin-oxc-transform": "file:plugins/obuild-plugin-oxc-transform-0.0.1.tgz",
+    "@obuild/plugin-oxc-dts": "file:plugins/obuild-plugin-oxc-dts-0.0.1.tgz",
+    "@obuild/plugin-oxc-minify": "file:plugins/obuild-plugin-oxc-minify-0.0.1.tgz",
```
These should be built-in: (as of current version)
- plugin core
- oxc transformer, minifier, and dts generator

pi0 commented on src/builders/bundle.ts:
```diff
@@ -1,6 +1,6 @@
 import { builtinModules } from "node:module";
 import { mkdir, readFile, writeFile } from "node:fs/promises";
-import { dirname, relative, join, basename, extname, resolve } from "node:path";
+import { dirname, relative, join, basename, extname, resolve } from "pathe";
```
Do we really need `pathe`?

zsilbi replied:
In #43 it was reported that it helps with separator issueson Windows

pi0 commented on src/builders/transform/plugins.ts:
```diff
@@ -0,0 +1,74 @@
+  "tsgo-dts":            "@obuild/plugin-tsgo-dts",
+  "vue-sfc-transformer": "@obuild/plugin-vue-sfc-transformer",
+  "vue-tsc-dts":         "@obuild/plugin-vue-tsc-dts",
+};
```
We can try `{name}` => `@obuild/plugin-${name}` instead of keeping a hardcoded list

zsilbi committed:
feat: remove `pkg-types`

zsilbi committed:
chore: remove tsbuildinfo

zsilbi committed:
feat: add config options for built-in plugins

zsilbi committed:
chore: update lockfile

zsilbi committed:
docs: update readme

zsilbi committed:
chore: update plugins

zsilbi committed:
test: add tests for default plugins

zsilbi committed:
test: transformer and utils

zsilbi committed:
test: simplify mocks

zsilbi committed:
fix: raw files handling w/o srcPath

zsilbi committed:
feat: add plugin types

zsilbi committed:
chore: update plugins

zsilbi committed:
feat: dynamicly import builders based on entry type

zsilbi committed:
chore: use released `@obuild` packages
